Bluetooth: Controller: Add support for Extended Active Scanning

cvinayak opened the pull request:
Add support for Extended Active Scanning and generation of
Extended Advertising Report with Scan Response event type.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: Controller: nRF53x: Fix missing sw_switch clear DPPI config

Fix missing sw_switch timer clear DPPI config when
re-enabling Tx or Rx after radio_disable() or
radio_switch_complete_and_disable() call in LLL state/role
contexts.

Relates to commit 1c3659519f69 ("Bluetooth: controller:
nRF53x: Fix missing sw_switch clear DPPI config").

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: controller: Extended scanning with LLL scheduling

Extended scanning with LLL scheduling.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>
Signed-off-by: Andrzej Kaczmarek <andrzej.kaczmarek@codecoup.pl>

andrzej-kaczmarek committed:
Bluetooth: controller: Fix isr_done_cleanup

We need to reset isr status before returning as otherwise irq will be
triggered over and over again.

Signed-off-by: Andrzej Kaczmarek <andrzej.kaczmarek@codecoup.pl>
Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

andrzej-kaczmarek committed:
Bluetooth: controller: Add ull_scan_aux_is_valid_get helper

This can be used to check if pointer is a valid lll_scan_aux pointer,
e.g. if we need to decide what kind of struct is attached to RX node
for further processing.

Signed-off-by: Andrzej Kaczmarek <andrzej.kaczmarek@codecoup.pl>

cvinayak committed:
Bluetooth: Add define for Extended Scannable Advertising parameters

Add a helper define for Extended Scannable Advertising
Parameters.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: Controller: Add support for Extended Active Scanning

Add support for Extended Active Scanning and generation of
Extended Advertising Report with Scan Response event type.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
tests: Bluetooth: bsim: Verify Extended Active Scanning

Added tests to verify Extended Active Scanning support.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: Controller: Flush scanned rx node before scan response PDU

Flush the scanned advertising PDUs to HCI while receiving
the scan response PDU when Extended Active scanning.
This reverts earlier implementation to remember the
advertising PDUs.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: Controller: Hold back Extended Scannable PDUs

When Extended Active Scanning hold back the received PDUs in
a linked list and send it to HCI together so that fields
present across these PDUs can be aggregated for generating
the LE Extended Advertising Report.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: Controller: Generate Extended Scan Response report

Added implementation to traverse the Extended Active Scanned
advertising PDUs, generate the Scannable Advertising and
Scan Response Advertising Report.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: Controller: Flush Aux PDUs being in LLL scheduling

When Auxiliary PDUs are scanning by the primary scan's LLL
scheduling, then flush the Auxiliary PDUs when processing
them in the ULL instead of using the disabled callback.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: Controller: Resume primary PDU scanning after Auxiliary PDU

Add implementation to resume primary PDU scanning window
after use of LLL scheduling to receive auxiliary PDUs.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: Controller: Fix tag name should be unique violation

Fix tag name should be unique violation.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: Controller: LLL scheduling for active scanning and initiator

Implementation to use LLL scheduling for close auxiliary
offsets when active scanning and initiating connections.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: Controller: Simplify LLL scheduling decision

Simplify the implementation deciding the use of LLL
scheduling for scanning PDUs with close auxiliary offsets.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

Thalley commented on include/bluetooth/bluetooth.h:
```diff
@@ -704,6 +704,14 @@ struct bt_le_per_adv_param {
 						NULL)
 
+/** Scannable extended advertising with @ref BT_LE_ADV_OPT_USE_NAME */
+#define BT_LE_EXT_ADV_SCAN_NAME BT_LE_ADV_PARAM(BT_LE_ADV_OPT_EXT_ADV | \
+						BT_LE_ADV_OPT_SCANNABLE | \
+						BT_LE_ADV_OPT_USE_NAME, \
```
I was wondering why the other default parameters didn't have `BT_LE_ADV_OPT_SCANNABLE` when they are putting the device name in the scan response. From what I can tell from the stack code, `BT_LE_ADV_OPT_SCANNABLE` is implicitly set when `BT_LE_ADV_OPT_USE_NAME` is set, or if any scan data was provided in which doesn't seem to be documented at all. It even appears that if any `sd` is provided in `bt_le_adv_start` then `BT_LE_ADV_OPT_SCANNABLE` is also implicitly enabled. 

Is this by design and intention @joerchan ?

Thalley replied:
That is correct w.r.t. `sd` not being there anymore. However, the expression that adds scan data uses a boolean based on `sd || (param->options & BT_LE_ADV_OPT_USE_NAME)`, so even with the extended advertising API, if `BT_LE_ADV_OPT_USE_NAME` is set, then it is scannable. 

I actually see that it is not taking the `BT_LE_ADV_OPT_FORCE_NAME_IN_AD` option into account for that, so I have also created this issue: https://github.com/zephyrproject-rtos/zephyr/issues/37462

cvinayak replied:
> the expression that adds scan data uses a boolean based on sd || (param->options & BT_LE_ADV_OPT_USE_NAME)

Correct me, that piece of code path is only applicable for legacy advertising... `bt_le_ext_adv_create` always sets `false` to `has_scan_data` parameter when calling `le_ext_adv_param_set`

joerchan replied:
> Is this by design and intention @joerchan ?

Yes, ADV_OPT_SCANNABLE was added together with the extended advertising API, and it is not required for the legacy API in order to make the advertiser scannable. Setting the SD data (or OPT_NAME) makes it scannable implicitly. This maintains the old behavior and application did not have to update.

Thalley replied:
@cvinayak The check also appears in `bt_le_adv_start_ext`:
```
	err = le_ext_adv_param_set(adv, param, sd ||
				   (param->options & BT_LE_ADV_OPT_USE_NAME));
```

@joerchan Is this perhaps something that should be made clear in the API description?

Thalley commented on subsys/bluetooth/controller/hci/hci.c:
```diff
@@ -5100,11 +5104,16 @@ static void le_ext_adv_report(struct pdu_data *pdu_data,
 			sec_phy = sec_phy_curr;
 			data_len = data_len_curr;
 			total_data_len = data_len;
+			total_scan_data_len = 0U;
```
Possibly outside the scope of this PR: The do-while loop where this scan data is retrieved is now more than 200 lines of code, inside of a function that is now almost 400 lines of code. Is there any way we can split this into smaller blocks of code without any significant negative changes in performance or memory usage?

Thalley commented on subsys/bluetooth/controller/hci/hci.c:
```diff
@@ -5253,6 +5271,91 @@ static void le_ext_adv_report(struct pdu_data *pdu_data,
+	/* Allocate, append as buf fragement and construct the scan response
+	 * event.
+	 */
+	struct net_buf *scan_evt_buf = bt_buf_get_rx(BT_BUF_EVT, K_FOREVER);
```
Declare at top of scope?

Thalley commented on subsys/bluetooth/controller/hci/hci.c:
```diff
@@ -5253,6 +5271,91 @@ static void le_ext_adv_report(struct pdu_data *pdu_data,
+		return;
+	}
+
+	/* Allocate, append as buf fragement and construct the scan response
+	 * event.
+	 */
```
As per my previous comment: It seems like the following addition of scan response handling could be put into a separate function.

cvinayak replied:
Refactored in to `ext_adv_info_fill()`

Thalley commented on subsys/bluetooth/controller/hci/hci.c:
```diff
@@ -5253,6 +5271,91 @@ static void le_ext_adv_report(struct pdu_data *pdu_data,
+
+	net_buf_frag_add(buf, scan_evt_buf);
+
+	/* If scan data complete */
+	if (!scan_data_status) {
```
The following code seem very similar to exiting code for handling advertising reports (which are, of course, also sent with the same HCI event as scan responses). Is it much effort to adjust the current code for advertising reports to reuse the same code for scan responses, as basically the only difference is the event type? (at least from a HCI point of view).

cvinayak replied:
Refactored in to ext_adv_info_fill()

Thalley commented on subsys/bluetooth/controller/hci/hci.c:
```diff
@@ -5253,6 +5271,91 @@ static void le_ext_adv_report(struct pdu_data *pdu_data,
+
+	if (evt_type & BT_HCI_LE_ADV_EVT_TYPE_DIRECT) {
+		adv_info->direct_addr.type = direct_addr_type;
+		memcpy(&adv_info->direct_addr.a.val[0], direct_addr,
+		       sizeof(bt_addr_t));
```
Are you typically casting unused return values to `void` in the controller? If so, consider doing so for `memcpy` and `memset` as well (i.e. `(void)memcpy(...)`.

Thalley commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_scan_aux.c:
```diff
@@ -174,7 +178,7 @@ static int prepare_cb(struct lll_prepare_param *p)
 	/* setup tIFS switching */
 	radio_tmr_tifs_set(EVENT_IFS_US);
-	radio_switch_complete_and_tx(lll->phy, 0, lll->phy, 1);
+	radio_switch_complete_and_tx(lll->phy, 0, lll->phy, 0);
```
Since we haven't been using proper names for the flags in the code, can you elaborate on why we are changing the tx flag from 1 to 0 in a PR that deals with scanning?

Thalley commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_scan_aux.c:
```diff
@@ -584,27 +587,143 @@ static int isr_rx_pdu(struct lll_scan_aux *lll, uint8_t devmatch_ok,
+		lrpa = ull_filter_lll_lrpa_get(rl_idx);
+		if (lll_scan->rpa_gen && lrpa) {
+			pdu_tx->tx_addr = 1;
+			memcpy(&pdu_tx->scan_req.scan_addr[0], lrpa->val,
```
I might have asked this question before, but why are we doing `&pdu_tx->scan_req.scan_addr[0]` instead of `pdu_tx->scan_req.scan_addr`? Wouldn't that give the same result?

Thalley replied:
This comment applies to multiple places.

Thalley commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_scan_aux.c:
```diff
@@ -584,27 +587,143 @@ static int isr_rx_pdu(struct lll_scan_aux *lll, uint8_t devmatch_ok,
+		/* assert if radio packet ptr is not set and radio started tx */
+		LL_ASSERT(!radio_is_ready());
+
+		if (IS_ENABLED(CONFIG_BT_CTLR_PROFILE_ISR)) {
+			lll_prof_cputime_capture();
+
+		}
```
```suggestion
		if (IS_ENABLED(CONFIG_BT_CTLR_PROFILE_ISR)) {
			lll_prof_cputime_capture();
		}
```

Thalley commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_scan_aux.c:
```diff
@@ -584,27 +587,143 @@ static int isr_rx_pdu(struct lll_scan_aux *lll, uint8_t devmatch_ok,
+#else
+		if (1) {
+#endif /* CONFIG_BT_CTLR_PRIVACY */
+			pdu_tx->tx_addr = lll_scan->init_addr_type;
```
I though that the `tx_addr` was a bit to tell whether the address was set or not, but here you are assigning an address type to it. Is this intentional?

Thalley replied:
Wouldn't it be better to call it `tx_addr_type` then, like most of the variables that it is assigned to?

cvinayak replied:
Link Layer sections of the Spec does not use `type` word associated with `TxAdd` or `RxAdd`, but the HCI commands name the parameters with `_type`. Will continue to keep it this way to be consistent with the BT Spec.

Thalley commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_scan_aux.c:
```diff
@@ -584,27 +587,143 @@ static int isr_rx_pdu(struct lll_scan_aux *lll, uint8_t devmatch_ok,
+		radio_gpio_pa_setup();
+		radio_gpio_pa_lna_enable(radio_tmr_tifs_base_get() +
+					 EVENT_IFS_US -
+					 radio_rx_chain_delay_get(lll->phy, 1) -
```
The value `1` here is a flag, but what semantic value does it have? Do we have any `#define` that can be used here?

Thalley commented on subsys/bluetooth/controller/ll_sw/ull_scan_aux.c:
```diff
@@ -144,6 +146,8 @@ void ull_scan_aux_setup(memq_link_t *link, struct node_rx_hdr *rx)
 				return;
 			}
 
+			is_scan_req = !!ftr->scan_req;
+
```
Could this be moved to line 202 instead? Seems to make more sense to update when it needs to be used.

cvinayak committed:
Bluetooth: Controller: Release LLL scheduling aux on incomplete data

Add implementation to release Auxiliary Context on failure
to receive auxiliary chain being in LLL scheduling.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

andrzej-kaczmarek committed:
Bluetooth: Controller: Fix aux scheduling from LLL

We need to alloc node prior to calling lll_scan_aux_setup, otherwise
the same node will be peeked inside that function and it will overwrite
some data in our node (e.g. "extra" field).

Signed-off-by: Andrzej Kaczmarek <andrzej.kaczmarek@codecoup.pl>
Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

Thalley commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_scan_aux.c:
```diff
@@ -584,27 +587,142 @@ static int isr_rx_pdu(struct lll_scan_aux *lll, uint8_t devmatch_ok,
+		if (1) {
+#endif /* CONFIG_BT_CTLR_PRIVACY */
+			pdu_tx->tx_addr = lll_scan->init_addr_type;
+			memcpy(pdu_tx->scan_req.scan_addr,
```
Since you already started based on my previous comment ;) (applies multiple places). 

```suggestion
			(void)memcpy(pdu_tx->scan_req.scan_addr,
```

Thalley commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_scan_aux.c:
```diff
@@ -584,27 +587,142 @@ static int isr_rx_pdu(struct lll_scan_aux *lll, uint8_t devmatch_ok,
+
 		trx_cnt++;
 
+		ull_pdu_rx_alloc();
```
This can't fail since you have done the peak above, right? 

```suggestion
		(void)ull_pdu_rx_alloc();
```

Thalley reviewed (approved):
A few minor comments left, but none of them really needs to be addressed.

cvinayak committed:
Bluetooth: Controller: Reverting Extended scanning with LLL scheduling

Reverting Extended scanning with LLL scheduling as this is
causing unresolvable merge conflict with the commits that
is implementing the Extended Active Scanning support.

Subsequent commit add the revert implementation back again.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak commented:
@Thalley , @ppryga  and @andrzej-kaczmarek , I had to revert the commits in https://github.com/zephyrproject-rtos/zephyr/pull/36588 in this PR and add it back as I had unsolvable conflicts that broken functionally.

Please re-review the PR as a whole again.

cvinayak commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_scan_aux.c:
```diff
@@ -51,20 +51,22 @@ static int prepare_cb(struct lll_prepare_param *p);
 		      uint8_t irkmatch_id, uint8_t rl_idx, uint8_t rssi_ready);
-static void isr_rx_ull_scheduled(void *param);
-static void isr_rx_lll_scheduled(void *param);
+static void isr_tx_scan_req_ull_scheduled(void *param);
+static void isr_tx_scan_req_lll_scheduled(void *param);
```
Done

cvinayak commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_scan_aux.c:
```diff
@@ -431,6 +439,157 @@ static void isr_done(void *param)
+	lll_isr_status_reset();
+
+	node_rx = param;
+	ftr = &(node_rx->hdr.rx_ftr);
```
Done

cvinayak commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_scan_aux.c:
```diff
@@ -533,6 +703,7 @@ static int isr_rx_pdu(struct lll_scan *lll, struct lll_scan_aux *lll_aux,
 		   (pdu->adv_ext_ind.adv_mode & BT_HCI_LE_ADV_PROP_CONN) &&
 		   lll_scan_ext_tgta_check(lll, false, true, pdu,
 					   rl_idx)) {
+		struct lll_scan_aux *lll_lll_aux;
```
changed to `lll_aux_to_use`

cvinayak commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_scan_aux.c:
```diff
@@ -627,7 +800,7 @@ static int isr_rx_pdu(struct lll_scan *lll, struct lll_scan_aux *lll_aux,
 		radio_tmr_tifs_set(EVENT_IFS_US);
 		radio_switch_complete_and_rx(phy_aux);
-		radio_isr_set(isr_tx_connect_req, lll_aux);
+		radio_isr_set(isr_tx_connect_req, lll_lll_aux);
```
Changed to `lll_aux_to_use`.  The pointer is copied from `lll_aux` when ULL schedule is used from an explicitly allocated aux context, or supplied in `lll->lll_aux` when LLL schedule is used by the primary scan context which gets a deferred aux context allocated.

cvinayak committed:
Bluetooth: Controller: Fix scan_data_status value

Fix scan_data_status value from being incorrected OR-ed with
the advertising data status.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: Controller: Explicitly cast to void memq_dequeue calls

Explicitly cast to void memq_dequeue function calls.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: Controller: Extended Active Scanning review rework

Miscellaneous review rework changes.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
tests: Bluetooth: bsim: Add a little wait for connection establishment

Add a little wait before completing the test to ensure both
central and peripheral side of the test completes before the
test exits.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>
